python/friesian/example/multi_task/data_processing.py

```diff
@@ -0,0 +1,159 @@
+#
+# Copyright 2016 The BigDL Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import argparse
+import os
+from argparse import ArgumentParser
+
+from bigdl.friesian.feature import FeatureTable
+from bigdl.orca import init_orca_context, stop_orca_context
+
+
+def transform(x):
+    if x == '上海':
+        return 0.0
+    elif isinstance(x, float):
+        return float(x)
+    else:
+        return float(eval(x))
+
+
+def transform_cat_2(x):
+    return '-'.join(sorted(x.split('/')))
+
+
+def read_and_split(data_input_path, sparse_int_features, sparse_string_features, dense_features):
+    header_names = ['user_id', 'article_id', 'expo_time', 'net_status', 'flush_nums',
+                    'exop_position', 'click', 'duration', 'device', 'os', 'province', 'city',
+                    'age', 'gender', 'ctime', 'img_num', 'cat_1', 'cat_2'
+                    ]
+    if data_input_path.split('.')[-1] == 'csv':
+        data_pd = FeatureTable.read_csv(data_input_path, header=False, names=header_names)
+    else:
+        data_pd = FeatureTable.read_parquet(data_input_path)
+    data_pd = data_pd.cast(sparse_int_features, 'string')
+    data_pd = data_pd.cast(dense_features, 'string')
+
+    # fill absence data
+    for feature in (sparse_int_features + sparse_string_features):
+        data_pd = data_pd.fillna("", feature)
+    for dense_feature in dense_features:
+        data_pd = data_pd.fillna('0.0', dense_feature)
+    print(data_pd.df.dtypes)
+
+    process_img_num = lambda x: transform(x)
+    process_cat_2 = lambda x: transform_cat_2(x)
+    data_pd = data_pd.apply("img_num", "img_num", process_img_num, "float")
+    data_pd = data_pd.apply("cat_2", "cat_2", process_cat_2, "string")
+
+    train_tbl = FeatureTable(data_pd.df[data_pd.df['expo_time'] < '2021-07-06'])
+    valid_tbl = FeatureTable(data_pd.df[data_pd.df['expo_time'] >= '2021-07-06'])
+    print('train_data.shape: ', train_tbl.size())
+    print('test_data.shape: ', valid_tbl.size())
+    return train_tbl, valid_tbl
+
+
+def feature_engineering(train_tbl, valid_tbl, model_path, model_path_json, sparse_int_features,
+                        sparse_string_features, dense_features):
+    import json
+    train_tbl, min_max_dict = train_tbl.min_max_scale(dense_features)
+    valid_tbl = valid_tbl.transform_min_max_scale(dense_features, min_max_dict)
+    cat_cols = sparse_string_features[-1:] + sparse_int_features + sparse_string_features[:-1]
```

Comment:
is sparse_string_features[-1:] + sparse_string_features[:-1] meaningful?

Reply:
This is to avoid a strange bug, detailed information I previously sent you through teams on 8.18.